File: assignments/rust/nucleotide-count/nucleotide-count_test.rs
#[link(name = "dna", vers = "1.0")];
#[crate_type = "lib"];

use std::hashmap::HashMap;

mod dna;

fn check_dna(s: &str, pairs: &[(char, uint)]) {
    // The reason for the awkward code in here is to ensure that the failure
    // message for assert_eq! is as informative as possible. A simpler
    // solution would simply check the length of the map, and then
    // check for the presence and value of each key in the given pairs vector.
    let mut m: HashMap<char, uint> = dna::nucleotide_counts(s);
    for &(k, v) in pairs.iter() {
        assert_eq!((k, m.pop(&k).unwrap()), (k, v));
    }
    // may fail with a message that clearly shows all extra pairs in the map
    assert_eq!(m.iter().to_owned_vec(), ~[]);
}

#[test]
#[ignore]
fn test_count_empty() {
    assert_eq!(dna::count('A', ""), 0);
}

#[test]
#[ignore]
fn test_count_repetitive_cytidine() {
    assert_eq!(dna::count('C', "CCCCC"), 5);
}

#[test]
#[ignore]
fn test_count_only_thymidine() {
    assert_eq!(dna::count('T', "GGGGGTAACCCGG"), 1);
}

#[test]
#[ignore]
fn test_count_dna_has_no_no_uracil() {
    assert_eq!(dna::count('U', "GATTACA"), 0);
}

/*
testCase "validates nucleotides" $
assertError "invalid nucleotide 'X'" $ count 'X' "GACT"
*/

#[test]
#[ignore]
fn test_nucleotide_count_empty() {
    check_dna(
        "",
        [('A', 0), ('T', 0), ('C', 0), ('G', 0)]);
}

#[test]
#[ignore]
fn test_nucleotide_count_only_guanosine() {
    check_dna(
        "GGGGGGGG",
        [('A', 0), ('T', 0), ('C', 0), ('G', 8)]);
}

#[test]
#[ignore]
fn test_nucleotide_count_counts_all() {
    check_dna(
        "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAA\
         GAGTGTCTGATAGCAGC",
        [('A', 20), ('T', 21), ('C', 12), ('G', 17)]);
}

File: assignments/elixir/nucleotide-count/nucleotide_count_test.exs
Code.load_file("dna.exs")
ExUnit.start

defmodule DNATest do
  use ExUnit.Case, async: true
  doctest DNA

  test "empty dna string has no adenosine" do
    assert DNA.count('', ?A) == 0
  end

  test "empty dna string has no nucleotides" do
    expected = HashDict.new [{?A, 0}, {?T, 0}, {?C, 0}, {?G, 0}]
    assert expected == DNA.nucleotide_counts('')
  end

  test "repetitive cytidine gets counted" do
    assert 5 == DNA.count('CCCCC', ?C)
  end

  test "repetitive sequence has only guanosine" do
    expected = HashDict.new [{?A, 0}, {?T, 0}, {?C, 0}, {?G, 8}]
    assert expected == DNA.nucleotide_counts('GGGGGGGG')
  end

  test "counts only thymidine" do
    assert 1 == DNA.count('GGGGGTAACCCGG', ?T)
  end

  test "dna has no uracil" do
    assert 0 == DNA.count('GATTACA', ?U)
  end

  test "counts all nucleotides" do
    s = 'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'
    expected = HashDict.new [{?A, 20}, {?T, 21}, {?C, 12}, {?G, 17}]
    assert expected == DNA.nucleotide_counts(s)
  end
end
